remote_attestation/rust/src/attestation.rs

```diff
@@ -15,238 +15,545 @@
+                ephemeral_public_key,
+                random: self.state.random.to_vec(),
+                transcript_signature: vec![],
+                signing_public_key: vec![],
+                attestation_info: None,
+            }
+        };
+
+        let attestation_engine = self
+            .transition(key_negotiator)
+            .context("Couldn't transition from Initializing to Attesting state")?;
+        Ok((server_identity, attestation_engine))
+    }
+}
+
+impl AttestationEngine<Client, Attesting> {
+    /// Responds to `AttestationInit` message by creating a `ClientIdentity` message and derives
+    /// session keys for encrypting/decrypting messages from the server.
+    ///
+    /// `ClientIdentity` message contains an ephemeral public key for negotiating session keys.
+    /// If self attestation is enabled this message also provides necessary information to perform
+    /// remote attestation.
+    ///
+    /// Transitions [`AttestationEngine`] state from [`Attesting`] to [`Attested`] state.
+    pub fn process_server_identity(
+        mut self,
+        server_identity: &ServerIdentity,
+    ) -> anyhow::Result<(ClientIdentity, AttestationEngine<Client, Attested>)> {
+        if self.behavior.contains_peer_attestation() {
+            // Verify server transcript signature.
+            // Transcript doesn't include transcript signature from the server identity message.
+            let mut server_identity_no_signature = server_identity.clone();
+            server_identity_no_signature.transcript_signature = vec![];
+            self.transcript
+                .append(&server_identity_no_signature)
+                .context("Couldn't append server identity to transcript")?;
+            let verifier = SignatureVerifier::new(&server_identity.signing_public_key);
+            verifier
+                .verify(
+                    &self.transcript.get_sha256(),
+                    &server_identity.transcript_signature,
+                )
+                .context("Couldn't verify server transcript")?;
+
+            // Verify server attestation info.
+            let server_attestation_info = &server_identity
+                .attestation_info
+                .as_ref()
+                .context("Couldn't get server attestation info")?;
+            let expected_tee_measurement = self
+                .behavior
+                .get_expected_tee_measurement()
+                .as_ref()
+                .context("Couldn't get expected TEE measurement")?;
+            verify_attestation_info(&server_attestation_info, &expected_tee_measurement)
+                .context("Couldn't verify server attestation info")?;
+        }
+
+        // Create client identity message.
+        let ephemeral_public_key = self
+            .state
+            .key_negotiator
+            .public_key()
+            .context("Couldn't get ephemeral public key")?;
+        let client_identity = if self.behavior.contains_self_attestation() {
```

Comment:
Again could there be a method in `AttestationBehavior` for creating `client_identity`?